lib/execution_engine2/sdk/EE2Runjob.py

```diff
@@ -457,6 +498,137 @@ def _check_job_arguments(self, jobs, has_parent_job=False):
             self._check_ws_objects(source_objects=job.get(_SOURCE_WS_OBJECTS))
 
+    def retry_multiple(
+        self, job_ids, as_admin=False
+    ) -> List[Dict[str, Union[str, Any]]]:
+        """
+        #TODO Add new job requirements/cgroups as an optional param
+        :param job_ids: The list of jobs to retry
+        :param as_admin: Run with admin permission
+        :return: The child job ids that have been retried or errors
+        """
+        retried_jobs = []
+        for job_id in job_ids:
+            try:
+                retried_jobs.append(self.retry(job_id, as_admin=as_admin))
+            except Exception as e:
+                retried_jobs.append({"job_id": job_id, "error": f"{e}"})
+        return retried_jobs
+
+    @staticmethod
+    def _retryable(status: str):
+        return status in [Status.terminated.value, Status.error.value]
+
+    def retry(self, job_id: str, as_admin=False) -> Dict[str, Optional[str]]:
+        """
+        #TODO Add new job requirements/cgroups as an optional param
+        :param job_id: The main job to retry
+        :param as_admin: Run with admin permission
+        :return: The child job id that has been retried
+        """
+        job = self.sdkmr.get_job_with_permission(
+            job_id, JobPermissions.WRITE, as_admin=as_admin
+        )  # type: Job
+
+        # TODO: Fix this when updating the run_job_batch endpoint
+        # if job.batch_job:
+        #     raise CannotRetryJob("Cannot retry parent container jobs")
+
+        if not self._retryable(job.status):
+            raise CannotRetryJob(
+                f"Can only retry a cancelled or errored job_id:{job_id} status:{job.status}"
+            )
+
+        # Cannot retry a retried job, you must retry the parent
+        if job.retry_parent:
+            return self.retry(str(job.retry_parent), as_admin=as_admin)
+
+        # Get run job params from db, and inject parent job id, then run it
+        run_job_params = self._get_run_job_params_from_existing_job(
+            job, user_id=self.sdkmr.user_id
+        )
+
+        # Submit job to job scheduler or fail and not count it as a retry attempt
+        run_job_params[_PARENT_RETRY_JOB_ID] = job_id
+        retry_job_id = self.run(params=run_job_params, as_admin=as_admin)
+
+        # Save that the job has been retried, and increment the count. Notify the parent(s)
+        # 1) Notify the retry_parent that it has been retried
+        # TODO Use and create a method in sdkmr?
+        job.modify(inc__retry_count=1, set__retried=True)
+        # 2) Notify the parent container that it has a new child.. #TODO, maybe move this into multiple retries section
+        if run_job_params[_PARENT_JOB_ID]:
+            parent_job = self.sdkmr.get_job_with_permission(
+                run_job_params[_PARENT_JOB_ID], JobPermissions.WRITE, as_admin=as_admin
+            )
+            parent_job.modify(push__child_jobs=retry_job_id)
```

Comment:
Server / DB / network failures between
- running the retry job
- updating the retry parent
- updating the job parent

... could leave the DB in an inconsistent state.

Reply:
Ok I attempted to add some logic to handle a small subset of these cases